packages/block-library/src/navigation/index.php

```diff
@@ -454,13 +455,59 @@ function block_core_navigation_get_default_pages_fallback() {
 		true // So that we can check whether the result is an error.
 	);
 
-	if ( is_wp_error( $wp_insert_post_result ) ) {
+	return $wp_insert_post_result;
+}
+
+/**
+ * Creates an appropriate fallback Navigation Menu (`wp_navigation` Post)
+ * to be used by the Navigation block.
+ *
+ * Behaviour:
+ * 1. If there is a Navigation menu already then use it.
+ * 2. If there is a Classic menu then convert it to a Navigation menu.
+ * 3. If there is no Navigation menu and no Classic menu then create a default menu.
+ *
+ * The fallback menu is only created if the theme is a block theme.
+ *
+ * @return void
+ */
+function block_core_navigation_create_fallback() {
+	$should_skip = apply_filters( 'block_core_navigation_skip_fallback', false );
+
+	if ( $should_skip ) {
 		return;
 	}
 
-	// Fetch the most recently published navigation which will be the default one created above.
-	return block_core_navigation_get_most_recently_published_navigation();
+	// Don't create a fallback if the theme is not a block theme.
+	if ( ! wp_is_block_theme() ) {
+		return;
+	}
+
+	// Get the most recently published Navigation menu.
+	$existing_navigation_menu = block_core_navigation_get_most_recently_published_navigation();
+
+	// If there is already a Navigation menu then exit.
+	if ( $existing_navigation_menu ) {
+		return;
+	}
+
+	// If there are no Navigation menus then try to find a Classic menu
+	// and attempt to convert it into a Navigation menu.
+	$converted_classic_menu_id = block_core_navigation_create_classic_menu_fallback();
+
+	// If the conversion + creation was successful then exit.
+	if ( ! is_wp_error( $converted_classic_menu_id ) ) {
+		return;
+	}
+
+	// If the Classic menu process did not result in a post
+	// then default to creating a default fallback Navigaiton menu.
+	block_core_navigation_create_default_fallback();
 }
+// Run on switching Theme and when installing WP for the first time.
+add_action( 'switch_theme', 'block_core_navigation_create_fallback' );
+add_action( 'wp_install', 'block_core_navigation_create_fallback' );
```

Comment:
Discussing away from Github with @azaozz I'm concerned that hooking to `wp_install` might be a bad option because it's likely to mean _all_ sites will get a `wp_navigation` even if no block Theme is active.